seqr/views/utils/variant_utils.py

```diff
@@ -329,56 +329,36 @@ def get_saved_discovery_variants_by_family(variant_filter, format_variants, get_
-                unaffected_indivs_with_het_variants.add(sample_guid)
-
-    # AR-homozygote, AR-comphet, AR, AD, de novo, X-linked, UPD, other, multiple
-    if affected_indivs_with_hom_alt_variants:
-        if is_x_linked:
-            inheritance_models.add("X-linked")
-        else:
-            inheritance_models.add("AR-homozygote")
 
-    if not unaffected_indivs_with_het_variants and affected_indivs_with_het_variants:
-        inherited = any(
-            guid for guid in affected_indivs_with_het_variants
-            if any(parent_guid in affected_indivs_with_het_variants for parent_guid in parent_guid_map[guid])
-        )
-        if inherited or not unaffected_individual_guids:
-            inheritance_models.add("AD")
-        else:
-            inheritance_models.add("de novo")
+    is_x_linked = 'X' in variant_json.get('chrom', '')
+    genotype_zygosity = {
+        sample_guid: _get_genotype_zygosity(genotype, is_hemi_variant=is_x_linked and sample_guid in male_individual_guids)
+        for sample_guid, genotype in variant_json.get('genotypes', {}).items()
+    }
 
-    potential_compound_het_gene_ids = set()
-    if (len(unaffected_individual_guids) < 2 or unaffected_indivs_with_het_variants) \
-            and affected_indivs_with_het_variants and not affected_indivs_with_hom_alt_variants \
-            and 'transcripts' in variant_json:
-        potential_compound_het_gene_ids.update(list(variant_json['transcripts'].keys()))
+    affected_zygosities = {genotype_zygosity[g] for g in affected_individual_guids if g in genotype_zygosity}
+    unaffected_zygosities = {genotype_zygosity[g] for g in unaffected_individual_guids if g in genotype_zygosity}
 
-    return inheritance_models, potential_compound_het_gene_ids, genotype_zygosity
+    inheritance_model = None
+    potential_compound_het_gene_ids = set()
+    if any(zygosity in unaffected_zygosities for zygosity in {HOM_ALT, HEMI}):
+        # No valid inheritance modes for hom alt unaffected individuals
+        inheritance_model = None
+    elif any(zygosity in affected_zygosities for zygosity in {HOM_ALT, HEMI}):
+        inheritance_model = 'X-linked' if is_x_linked else 'AR-homozygote'
+    elif HET in affected_zygosities:
+        if HET not in unaffected_zygosities:
+            inherited = (not unaffected_individual_guids) or any(
+                guid for guid in affected_individual_guids
+                if genotype_zygosity.get(guid) == HET and
+                any(genotype_zygosity.get(parent_guid) == HET for parent_guid in parent_guid_map[guid])
+            )
+            inheritance_model = 'AD' if inherited else 'de novo'
+
+        if (len(unaffected_individual_guids) < 2 or HET in unaffected_zygosities) and 'transcripts' in variant_json:
+            potential_compound_het_gene_ids.update(list(variant_json['transcripts'].keys()))
+
+    return inheritance_model, potential_compound_het_gene_ids, genotype_zygosity
```

Comment:
Have you considered splitting this function that returns 3 values into separate functions that return 1, or returning a namedtuple or dictionary instead here? Then we'd never get the vague "ValueError: not enough values to unpack (expected 3, got 2)" again.

Reply:
I hear that, and I see how that could make things more resilient in some ways, but I would rather leave this function signature as is. I prefer the explicit return of the different things we expect instead of a dictionary with magic string keys, or the maintenance overhead of a named tuple thats making it seem like theres a real relationship between these parsed values when in fact they are just 3 different things that happen to be computed at the same time. This issue arose not because the calling functions were not expecting the right number of values, but because there was a branch in this fucntion that returned early on an edge case. To me, the solution is to have a single return statement per function, so in the future any changes to the signature only need one place to change instead of remembering to track down multiple possible return branches

Reply:
Actually, looking more into the usage, I think a dictionary would work better bcause I can clean up some of the logic that way, as these fields are used to update a dictionary in the end